src/util/address.rs

```diff
@@ -222,48 +363,88 @@ impl Display for Address {
                 };
                 prefixed[1..].copy_from_slice(&hash[..]);
                 base58::check_encode_slice_to_fmt(fmt, &prefixed[..])
-            },
-            Payload::WitnessProgram(ref witprog) => {
-                fmt.write_str(&witprog.to_address())
-            },
+            }
+            Payload::WitnessProgram {
+                version: ver,
+                program: ref prog,
+            } => {
+                let mut b32_data = vec![ver];
+                b32_data.extend_from_slice(&prog.to_base32());
+                let hrp = match self.network {
+                    Network::Bitcoin => "bc",
+                    Network::Testnet => "tb",
+                    Network::Regtest => "bcrt",
+                };
+                bech32::encode_to_fmt(fmt, &hrp, &b32_data).expect("only errors on invalid HRP")
+            }
         }
     }
 }
 
+/// Extract the bech32 prefix.
+/// Returns the same slice when no prefix is found.
+fn find_bech32_prefix(bech32: &str) -> &str {
```

Comment:
We should remember this code duplication when working on bech32 again. Maybe we can just export this function and avoid this redundancy.

Reply:
Yeah I thought the same. I didn't think it was worth it considering the size of the method.